### DISCUSSION TITLE
Cauchy Green Strain eigenvalues

### DISCUSSION THREAD

**Discussion by ZAARAOUI999** [adtzlr/matadi#135] [General]
Hi Mr Andreas, I wish you were fine, so I'm just asking if you have any idea about the results of the code given below. The ranking of the eigenvalues isn't the same, or it has no meaning?

The code:

```python

import numpy as np
from casadi import SX, Function
from matadi.math import eigvals

#
C = SX.sym('C',3,3)
#
_wc = eigvals(C)
wc = Function('wc',[C],[_wc])
#
F = np.eye(3)
F[0,0]= 0.25
F[0,2] = 0.5
F[2,1] = 0.3
#
C = F.T@F
#
λi_casadi = wc(C)
λi_numpy, _ = np.linalg.eig(C)
#
print('λi_casadi: ', λi_casadi)
print('λi_numpy: ', λi_numpy)

```

The results:
```
λi_casadi:  [1.48739, 0.0485864, 0.866525]
λi_numpy:  [0.0484878  1.48742393 0.86658827]
```
Thanks.

**Comment by adtzlr** [comment]
Hi, thanks, I'm fine! Hope you're fine too? It's because I added small numbers to the diagonal entries in the eig-function in matadi. If you use plain casadi, eigenvalues will be identical to numpy. This ensures a stable gradient in case of equal eigenvalues but introduces some (minimal) numeric errors.

Here's the code https://github.com/adtzlr/matadi/blob/6820ca54c8b9239acb0f7e11444e2207fdaca408/src/matadi/math.py#L158-L161

**Reply by adtzlr** [reply]
Did you encounter any problems related to that? Of course you could always set `eps=0` if you don't care about the gradients.

Personally, I found 1e-4 to be a good tradeoff between gradient stability and numeric errors.

**Reply by ZAARAOUI999** [reply]
I tried with the following code, but the ranking is not the same,

```python

import numpy as np
from casadi import SX, Function, eig_symbolic

#
C = SX.sym('C',3,3)
#
_wc = eig_symbolic(C)
wc = Function('wc',[C],[_wc])
#
F = np.eye(3)
F[0,0]= 0.25
F[0,2] = 0.5
F[2,1] = 0.3
#
C = F.T@F
#
λi_casadi = wc(C)
λi_numpy, _ = np.linalg.eig(C)
#
print('λi_casadi: ', λi_casadi)
print('λi_numpy: ', λi_numpy)
```

This can have an effect on some hyperelastic models that have a stretch-based form, taking for example a modified Yeoh model, where the strain energy is given by;
$$W =  \sum_{n=1}^3 C_{n0}(\lambda_n-1)^n$$
So here the ranking of eigenvalues should have sense.

**Reply by adtzlr** [reply]
Oh, you mean the order, not the small differences of the results. Sorry for my wrong interpretation of your question. Even numpy has different results on that, try to use `eig` and `eigh`. Isotropic hyperelasticity has always identical treatments of the stretches. Otherwise it's not isotropic. For non-isotropic stuff you have to sort the results manually - not every algorithm does that automatically.

See e.g. the numpy docs of `eig`: *...The eigenvalues are not necessarily ordered...*

**Reply by adtzlr** [reply]
Do you have a literature reference on your Yeoh equation? For me that formulation does not make sense, are you sure it is $\lambda_n$ and not $\lambda_\alpha$ with a second sum inside the brackets?

**Reply by ZAARAOUI999** [reply]
Thank you for your time Mr Andreas, I didn't know that before and I should read more about this valuable information ```Isotropic hyperelasticity has always identical treatments of the stretches.```.
All the best.

**Reply by adtzlr** [reply]
Never mind. To be fair, you haven't written about isotropy - I assumed that because you mentioned the Yeoh model formulation.

**Comment by ZAARAOUI999** [comment]
> Do you have a literature reference on your Yeoh equation? For me that formulation does not make sense, are you sure it is λn and not λα with a second sum inside the brackets?

No it was just an example where the invariants are replaced by the stretches.
